File: app/src/main/java/com/islam/basepropject/ui/ExPaging/ExDataSource.java
package com.islam.basepropject.ui.ExPaging;

import android.os.SystemClock;

import com.islam.basepropject.project_base.base.other.BaseDataSource;
import com.islam.basepropject.project_base.base.other.BaseViewModel;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Single;

public class ExDataSource extends BaseDataSource<String> {

    public ExDataSource(BaseViewModel baseView) {
        super(baseView);
    }

    int number = 1;
    @Override
    public Single<List<String>> fetchData(int key) {
        SystemClock.sleep(4000);
        List<String> strings = new ArrayList<>();
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        strings.add("string "+number++);
        return Single.just(strings) .subscribeOn(getBaseViewModel().getSchedulerProvider().io())
                .observeOn(getBaseViewModel().getSchedulerProvider().ui());
    }
}

File: app/src/main/java/com/islam/basepropject/ui/ExPaging/ExDataSource.kt
package com.islam.basepropject.ui.ExPaging

import com.islam.basepropject.pojo.Order
import com.islam.basepropject.project_base.base.other.BaseDataSource
import com.islam.basepropject.project_base.base.other.BaseViewModel
import com.islam.basepropject.project_base.views.OnViewStatusChange
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import java.util.*

class ExDataSource(baseView: BaseViewModel,val viewId: Int?) : BaseDataSource<Order>(baseView) {

    internal var number = 1
    override suspend fun fetchData(key: Int): List<Order> {
     val result =  baseViewModel.networkCall(viewId) {getData()}
        baseViewModel.markAsCompleted(listOf(result))
        return result.value!!
    }
    private suspend fun getData():List<Order>{
        return withContext(Dispatchers.IO) {
            delay(2500)
            val strings = ArrayList<Order>()
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings.add(Order("string " + number++))
            strings
        }
    }
}
